### PR TITLE
Require Institution Header to match institutionId of requests

### LINKED ISSUE #504: Check headers to ensure API calls refer to allowed institutions
The `CFPB-HMDA-Institutions` header will contain a comma-separated list of institution IDs (probably integers, see #437). We will check for its existence on all resources rooted from /institutions. Think of it as a whitelist of institutions authorized to a given user.

Return a 403 error if an attempt is made to access an institution-specific resource (/institutions/<institution_id>/*) not in the user's whitelist.

(Splitting this issue off from #441, since it requires more groundwork.)

### PR DESCRIPTION
**Pull request by schbetsy**
Closes #504 
Also includes minor cleanup of `InstitutionsHttpApiSpec`

As usual, I'm looking for feedback on my approach, idiomatic use of Scala, etc.

### CODE REVIEW

**Comment by schbetsy** [comment]
This will require the front-end to update their headers [here](https://github.com/cfpb/hmda-platform-ui/blob/master/src/js/api.js#L9). For testing, that line could be

```
'CFPB-HMDA-Institutions': '0,1,2 ,3'
```

**Comment by jmarin** [comment]
Overall, LGTM. One question though. Some of this code is using the Java API (i.e. [here](https://github.com/cfpb/hmda-platform/pull/555/files#diff-95ceb1992c3531f923df7c469b47af7aR62)). Have you considered implementing some of this logic by leveraging the built in header directives? (http://doc.akka.io/docs/akka/2.4.9/scala/http/routing-dsl/directives/header-directives/index.html)

**Comment by schbetsy** [comment]
I chose the `authorize` directive over the `header`-related directives when I originally implemented the header checking, but I don't remember exactly why. Thanks for the tip, I'll revisit that decision and see whether the built-in directives would be better.

File: api/src/main/scala/hmda/api/http/HmdaCustomDirectives.scala [hunk 1]
```diff
@@ -17,25 +16,32 @@ trait HmdaCustomDirectives extends ApiErrorProtocol {
     RejectionHandler.newBuilder()
       .handle {
         case AuthorizationFailedRejection =>
-          val errorResponse = ErrorResponse(403, "Unauthorized Access", "")
-          complete(ToResponseMarshallable(StatusCodes.Forbidden -> errorResponse))
+          extractUri { uri =>
+            val errorResponse = ErrorResponse(403, "Unauthorized Access", uri.toString)
```
Comment (hkeeler) [inline api/src/main/scala/hmda/api/http/HmdaCustomDirectives.scala]: This is going to result in the full URI being put into the `path` attribute, while the rest of the routes look like they're using a manually constructed path string, and passing that into all errors.  I think we should make this consistent.

I personally prefer having the full URI, and I also think we could replace all of the manual path building with a single directive or error handler.  That seems out outside the scope of this PR, though.  Perhaps for now, we just do a `uri.path.toString` for now to make it consistent?

Comment (schbetsy) [inline api/src/main/scala/hmda/api/http/HmdaCustomDirectives.scala]: +1 for consistency. Updated in 13c1c5f84c4903390f71c3c356f52900a6192178.

**Comment by hkeeler** [comment]
Looks good.  Merging...

**Pull request merged**